### PR TITLE
New algorithm for forward filling

### LINKED ISSUE #6112: Forward Fill not working when there are all-NaN chunks
**What happened**: I'm working with a report dataset that only has data on some specific periods of time, the problem is that when I use the forward fill method it returns me many nans even on the last cells (it's a forward fill without limit).

**What you expected to happen**: The array should not have nans in the last cells if it has data in any other cell or there should be a warning somewhere.

**Minimal Complete Verifiable Example**:

```python
import xarray as xr

xr.DataArray(
    [1, 2, np.nan, np.nan, np.nan, np.nan],
    dims=['a']
).chunk(
    2
).ffill(
    'a'
).compute()
```
output: array([ 1.,  2.,  2.,  2., nan, nan])

**Anything else we need to know?**: I check a little bit the internal code of Xarray for forward filling when it use dask and I think that the problem is that the algorithm makes an initial forward fill on all the blocks and then it makes a map_overlap for forward filling between chunks which in case that there is an empty chunk will not work due that it is going to take the last value of the empty chunk which is nan (hope this help).
**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.0 | packaged by conda-forge | (default, Nov 20 2021, 02:25:18) [GCC 9.4.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-1025-aws
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: None
libnetcdf: None

xarray: 0.20.2
pandas: 1.3.5
numpy: 1.21.4
scipy: None
netCDF4: None
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: 2.10.3
cftime: None
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: 1.3.2
dask: 2021.12.0
distributed: 2021.12.0
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
fsspec: 2021.11.1
cupy: None
pint: None
sparse: None
setuptools: 59.4.0
pip: 21.3.1
conda: None
pytest: 6.2.5
IPython: 7.30.1
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by josephnowak**
- [x] Closes #6112
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by josephnowak** [comment]
this fix #6112

File: xarray/tests/test_missing.py [hunk 1]
```diff
@@ -451,10 +451,6 @@ def test_ffill_bfill_dask(method):
     expected = numpy_method("time", limit=3)
     assert_equal(actual, expected)
 
-    # limit < axis size
-    with pytest.raises(NotImplementedError):
```
Comment (dcherian) [inline xarray/tests/test_missing.py:455]: if this works  now lets compare the result of `dask_method` with `numpy_method` like the other tests.

File: xarray/core/dask_array_ops.py [hunk 2]
```diff
@@ -57,24 +57,41 @@ def push(array, n, axis):
...
-            axis=axis,
-            n=n,
-            depth={axis: (1, 0)},
-            boundary="none",
-            dtype=array.dtype,
-            meta=array._meta,
+    def _fill_with_last_one(a, b):
+        # cumreduction apply the push func over all the blocks first so, the only missing part is filling
+        # the missing values using the last data of the previous chunk
+        if isinstance(a, np.ma.masked_array) or isinstance(b, np.ma.masked_array):
```
Comment (dcherian) [inline xarray/core/dask_array_ops.py:67]: I don't think masked arrays really work with xarray (could be wrong though), so we could just delete this bit.

**Review by dcherian — COMMENTED** [review]
Thanks @josephnowak Can  you add a note to `whats-new.rst` too please?

File: xarray/core/dask_array_ops.py (hunk 2 above)
Comment (josephnowak) [reply xarray/core/dask_array_ops.py:67]: No problem, I saw that logic on the dask methods of cumsum and cumprod so I thought that kind of cases were necessary

**Comment by josephnowak** [comment]
@dcherian I think that with the last changes everything is ready.

**Review by dcherian — COMMENTED** [review]
This looks great, and the test seem thorough to me.

Thanks  @josephnowak , this is a great first  PR.  I'll merge in  a couple of  days  if there are no more comments.

**Review by dcherian — APPROVED** [review]

**Comment by max-sixty** [comment]
Thanks @josephnowak !

**Pull request merged**